remotes/docker/fetcher.go

```diff
@@ -95,41 +95,51 @@ func (r dockerFetcher) Fetch(ctx context.Context, desc ocispec.Descriptor) (io.R
 			images.MediaTypeDockerSchema1Manifest,
 			ocispec.MediaTypeImageManifest, ocispec.MediaTypeImageIndex:
 
+			var lastErr error
 			for _, host := range r.hosts {
 				req := r.request(host, http.MethodGet, "manifests", desc.Digest.String())
 
 				rc, err := r.open(ctx, req, desc.MediaType, offset)
 				if err != nil {
-					if errdefs.IsNotFound(err) {
-						continue // try another host
+					// Store the error for referencing later
+					if lastErr == nil {
+						lastErr = err
 					}
-
-					return nil, err
+					continue // try another host
 				}
 
 				return rc, nil
 			}
+
+			if lastErr != nil {
+				return nil, lastErr
+			}
 		}
 
 		// Finally use blobs endpoints
+		var lastErr error
 		for _, host := range r.hosts {
 			req := r.request(host, http.MethodGet, "blobs", desc.Digest.String())
 
 			rc, err := r.open(ctx, req, desc.MediaType, offset)
 			if err != nil {
-				if errdefs.IsNotFound(err) {
-					continue // try another host
+				// Store the error for referencing later
+				if lastErr == nil {
```

Comment:
I guess there is kind of a weird condition that could happen here. Say you have a mirror + upstream, if the mirror has it not found then the upstream returns a connection error, what should the returned error be? That case could represent a case where the mirror simply could not mirror data from the upstream. I mentioned that comparing errors is hard, maybe it is best not to travel down that hole. wdyt?

Reply:
Yep definitely see where you are coming from - but might make things overly complicated if we tried to handle this case.